### PR TITLE
[Model] Add support for Qwen2MoeModel

### PR DESCRIPTION
**Pull request by wenyujin333**
Recently, the new codes of Qwen2MoE to Hugging Face Transformers, and thus I would love to contribute the new model to vLLM as well.

In this PR, I have provided the implementation of Qwen2MoE model and add some notes on it.

### CODE REVIEW

**Comment by JustinLin610** [comment]
@simon-mo hi, our codes are merged into HF transformers already. would you mind taking a look?

**Comment by WoosukKwon** [comment]
@esmeetu Could you take a look?

File: vllm/model_executor/models/qwen2_moe.py [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+from vllm.model_executor.parallel_utils.communication_op import (
+    tensor_model_parallel_all_reduce)
+from vllm.model_executor.parallel_utils.parallel_state import (
+    get_tensor_model_parallel_rank, get_tensor_model_parallel_world_size)
+from vllm.model_executor.sampling_metadata import SamplingMetadata
+from vllm.model_executor.weight_utils import (default_weight_loader,
+                                              hf_model_weights_iterator)
+from vllm.sequence import SamplerOutput
+
+KVCache = Tuple[torch.Tensor, torch.Tensor]
```
Comment (esmeetu) [inline vllm/model_executor/models/qwen2_moe.py:55]: This type is needless.

**Review by esmeetu — APPROVED** [review]
Left minor comments. And this look good to me.

File: vllm/model_executor/models/qwen2_moe.py (hunk 1 above)
Comment (wenyujin333) [reply vllm/model_executor/models/qwen2_moe.py]: already remove it

**Comment by wenyujin333** [comment]
@esmeetu it seems some unrelated tests failed, should we rerun it? or could you just merge this directly?

**Pull request merged**